File: src/android/java/com/hydra/noods/NooButton.java
/*
    Copyright 2019-2024 Hydr8gon

    This file is part of NooDS.

    NooDS is free software: you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    NooDS is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with NooDS. If not, see <https://www.gnu.org/licenses/>.
*/

package com.hydra.noods;

import android.content.Context;
import android.os.Vibrator;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup.LayoutParams;
import android.widget.Button;

public class NooButton extends Button
{
    private Vibrator vibrator;
    private int[] resIds;
    private int[] btnIds;
    private int state = 0;
    private int lastState = 0;

    public static int[] resAbxy =
    {
        R.drawable.abxy,          R.drawable.abxy_pressed1, R.drawable.abxy_pressed5, R.drawable.abxy,
        R.drawable.abxy_pressed7, R.drawable.abxy_pressed8, R.drawable.abxy_pressed6, R.drawable.abxy,
        R.drawable.abxy_pressed3, R.drawable.abxy_pressed2, R.drawable.abxy_pressed4, R.drawable.abxy,
        R.drawable.abxy,          R.drawable.abxy,          R.drawable.abxy,          R.drawable.abxy
    };

    public static int[] resDpad =
    {
        R.drawable.dpad,          R.drawable.dpad_pressed1, R.drawable.dpad_pressed5, R.drawable.dpad,
        R.drawable.dpad_pressed7, R.drawable.dpad_pressed8, R.drawable.dpad_pressed6, R.drawable.dpad,
        R.drawable.dpad_pressed3, R.drawable.dpad_pressed2, R.drawable.dpad_pressed4, R.drawable.dpad,
        R.drawable.dpad,          R.drawable.dpad,          R.drawable.dpad,          R.drawable.dpad
    };

    public NooButton(Context context, int[] resIds, int[] btnIds, int x, int y, int width, int height)
    {
        super(context);
        vibrator = (Vibrator)context.getSystemService(Context.VIBRATOR_SERVICE);
        this.resIds = resIds;
        this.btnIds = btnIds;

        // Set the button parameters
        setBackgroundResource(resIds[0]);
        setX(x);
        setY(y);
        setLayoutParams(new LayoutParams(width, height));
        setAlpha(0.5f);

        // Handle button touches
        if (btnIds.length >= 4) // D-pad
        {
            setOnTouchListener(new Button.OnTouchListener()
            {
                @Override
                public boolean onTouch(View view, MotionEvent event)
                {
                    switch (event.getAction())
                    {
                        case MotionEvent.ACTION_DOWN:
                        case MotionEvent.ACTION_MOVE:
                            // Press the right key if in range, otherwise release
                            if (event.getX() > view.getWidth() * 2 / 3)
                                pressDKey(0);
                            else
                                releaseKey(btnIds[0]);

                            // Press the left key if in range, otherwise release
                            if (event.getX() < view.getWidth() * 1 / 3)
                                pressDKey(1);
                            else
                                releaseKey(btnIds[1]);

                            // Press the up key if in range, otherwise release
                            if (event.getY() < view.getHeight() * 1 / 3)
                                pressDKey(2);
                            else
                                releaseKey(btnIds[2]);

                            // Press the down key if in range, otherwise release
                            if (event.getY() > view.getHeight() * 2 / 3)
                                pressDKey(3);
                            else
                                releaseKey(btnIds[3]);

                            // Update the image and vibrate
                            if (state != lastState)
                                setBackgroundResource(resIds[state]);
                            if ((state & ~lastState) != 0)
                                vibrator.vibrate(SettingsMenu.getVibrateStrength() * 4);
                            lastState = state;
                            state = 0;
                            break;

                        case MotionEvent.ACTION_UP:
                            // Release all directions and update the image
                            for (int i = 0; i < 4; i++)
                                releaseKey(btnIds[i]);
                            setBackgroundResource(resIds[0]);
                            lastState = state = 0;
                            break;
                    }
                    return true;
                }
            });
        }
        else
        {
            setOnTouchListener(new Button.OnTouchListener()
            {
                @Override
                public boolean onTouch(View view, MotionEvent event)
                {
                    switch (event.getAction())
                    {
                        case MotionEvent.ACTION_DOWN:
                            // Press a key, update the image, and vibrate
                            pressKey(btnIds[0]);
                            setBackgroundResource(resIds[1]);
                            vibrator.vibrate(SettingsMenu.getVibrateStrength() * 4);
                            break;

                        case MotionEvent.ACTION_UP:
                            // Release a key and update the image
                            releaseKey(btnIds[0]);
                            setBackgroundResource(resIds[0]);
                            break;
                    }
                    return true;
                }
            });
        }
    }

    private void pressDKey(int key)
    {
        // Press a D-pad key and track the state
        pressKey(btnIds[key]);
        state |= 1 << key;
    }

    public static native void pressKey(int key);
    public static native void releaseKey(int key);
}

File: src/ui/ios/NooButton.swift
/*
    Copyright 2019-2026 Hydr8gon

    This file is part of NooDS.

    NooDS is free software: you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    NooDS is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with NooDS. If not, see <https://www.gnu.org/licenses/>.
*/

import SwiftUI

struct NooButton: View {
    let ids: [Int32]
    var images = [Image]()
    @Binding private var menu: Bool
    @State private var state = 0

    init(menu: Binding<Bool>, ids: [Int32], name: String) {
        // Set values and add the button's released image
        self._menu = menu
        self.ids = ids
        images.append(Image(name).resizable())

        // Add the button's pressed images depending on type
        if ids.count == 1 { // Normal
            images.append(Image(name + "Pressed").resizable())
        }
        else { // D-pad
            for i in 1...8 {
                images.append(Image(name + "Pressed\(i)").resizable())
            }
        }
    }

    var body: some View {
        // Show an image selected by the current button state
        GeometryReader { proxy in
            images[state].opacity(0.5).simultaneousGesture(DragGesture(minimumDistance: 0)
                .onChanged({ touch in
                    // For normal buttons, check the ID for behavior
                    if ids.count == 1 {
                        if state != 1 && !menu {
                            if ids[0] == 12 {
                                // Open the in-game menu
                                menu = true
                            }
                            else {
                                // Change to pressed state and notify the core
                                state = 1
                                CoreBridge.pressKey(ids[0])
                            }
                            vibrate()
                        }
                        return
                    }

                    // For D-pads, check each direction and map the bitmask to an index
                    let old = state
                    state = 0
                    checkDir(cond: touch.location.x > proxy.size.width * 2 / 3, i: 0) // Right
                    checkDir(cond: touch.location.x < proxy.size.width * 1 / 3, i: 1) // Left
                    checkDir(cond: touch.location.y < proxy.size.height * 1 / 3, i: 2) // Up
                    checkDir(cond: touch.location.y > proxy.size.height * 2 / 3, i: 3) // Down
                    state = ([Int])([0, 1, 5, 0, 7, 8, 6, 0, 3, 2, 4])[state]
                    if state != old {
                        vibrate()
                    }
                })
                .onEnded({ _ in
                    // Change to released state and release all keys
                    state = 0
                    for id in ids {
                        CoreBridge.releaseKey(id)
                    }
                    vibrate()
                })
            )
        }
    }

    private func vibrate() {
        // Vibrate with the configured strength for button presses/releases
        if CoreBridge.vibrateStrength > 0 {
            let styles = ([UIImpactFeedbackGenerator.FeedbackStyle])([.soft, .light, .medium, .heavy, .rigid])
            UIImpactFeedbackGenerator(style: styles[Int(CoreBridge.vibrateStrength - 1)]).impactOccurred()
        }
    }

    private func checkDir(cond: Bool, i: Int) -> Void {
        // Update a direction's state and press/release its key
        if cond {
            state |= 1 << i
            CoreBridge.pressKey(ids[i])
        }
        else {
            CoreBridge.releaseKey(ids[i])
        }
    }
}
